<div align="center">
<h1>Jurnal Praktikum Dasar Kecerdasan Artifisial</h1>
<h3>Modul 2: Pengenalan Library NumPy</h3>
</div>

### Tujuan Praktikum
- Memahami konsep dasar array multidimensi dan bagaimana NumPy dapat digunakan untuk memanipulasi data secara efisien.
- Menguasai penggunaan fungsi dan operasi dasar NumPy, seperti pembuatan array, reshaping, indexing, serta pengolahan statistik.
- Mengaplikasikan NumPy dalam pemecahan masalah numerik dan pemodelan data untuk keperluan kecerdasan buatan dan analisis data.

In [1]:
print("'Aarif Rahmaan Jalaluddin Faqiih") # Tuliskan nama lengkap anda
print("103112430182") # Tuliskan nim anda
# print("") # Tuliskan nomor meja
print("PS1IF-12-REG06") # Tuliskan kelas anda
# print("") # Tuliskan kode asprak anda

'Aarif Rahmaan Jalaluddin Faqiih
103112430182
PS1IF-12-REG06


**Jangan lupa untuk selalu import Library NumPy memulai sesi baru (run time) atau script Python yang memerlukan penggunaan library NumPy.**

In [2]:
import numpy as np

### 1. Definisi Array dan Array NumPy (`5`)

**a.** Lengkapi potongan kode di bawah ini untuk membuat array Python biasa dan array NumPy, kemudian bandingkan perbedaan bentuknya.

In [3]:
# Inisialisasi array Python biasa
array_python = [103, 112, 430, 182]

# Inisialisasi array NumPy
array_numpy = np.array([103, 112, 430, 182])

# Cek tipe variabel
print("Tipe array_python:", type(array_python))
print("Tipe array_numpy:", type(array_numpy))

Tipe array_python: <class 'list'>
Tipe array_numpy: <class 'numpy.ndarray'>


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |'Tipe array_python: <class 'list'>'<br>'Tipe array_numpy: <class 'numpy.ndarray'>'|

**b.** Buat program yang menerima input dari pengguna berupa daftar bilangan, kemudian bentuk array Python biasa dan array NumPy dari daftar tersebut.

In [4]:
# Input dari user
data = [int(x) for x in input("Masukkan daftar bilangan, pisahkan dengan koma: ").split(",")]

# Masukkan data ke array Python biasa
array_python = data

# Masukkan data ke array NumPy
array_numpy = np.array(data)

# Tampilkan hasil
print("Array Python:", array_python)
print("Array NumPy:", array_numpy)

Masukkan daftar bilangan, pisahkan dengan koma: 1,2,3,4
Array Python: [1, 2, 3, 4]
Array NumPy: [1 2 3 4]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|1,2,3,4|Array Python: [1, 2, 3, 4]<br>Array NumPy: [1 2 3 4]|
|10,20,30|Array Python: [10, 20, 30]<br>Array NumPy: [10 20 30]|

### 2. Atribut Array NumPy (`10`)

Lengkapilah kode berikut untuk memeriksa atribut-atribut dimensi, bentuk, ukuran, dan tipe data dari array NumPy yang telah dibuat. Tujuan dari tugas ini adalah memahami bagaimana mendapatkan dimensi, bentuk, ukuran, dan tipe data dari array NumPy.

In [5]:
# Inisialisasi array NumPy
array_numpy = np.array([[1, 2, 3], [4, 5, 6]])

# Periksa jumlah dimensi
dimensi = array_numpy.ndim

# Periksa bentuk array (shape)
bentuk  = array_numpy.shape

# Periksa ukuran array (size)
ukuran  = array_numpy.size

# Periksa tipe data array (dtype)
tipe    = array_numpy.dtype

print("Jumlah dimensi:", dimensi)
print("Bentuk array:", bentuk)
print("Ukuran array:", ukuran)
print("Tipe data array:", tipe)

Jumlah dimensi: 2
Bentuk array: (2, 3)
Ukuran array: 6
Tipe data array: int64


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |'Jumlah dimensi: 2'<br>'Bentuk array: (2, 3)'<br>'Ukuran array: 6'<br>'Tipe data array: int64'|

### 3. Deklarasi Array dengan NumPy (`10`)

**a.** Inisialisasikan sebuah array 2D 2x3 yang berisi nilai nol.

In [6]:
arr = np.zeros((2, 3))
print(arr)

[[0. 0. 0.]
 [0. 0. 0.]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[0. 0. 0.]<br>[0. 0. 0.]]|

**b.** Inisialisasikan sebuah array 3D 3x2x4 yang berisi nilai satu.

In [7]:
arr = np.ones((3, 2, 4))
print(arr)

[[[1. 1. 1. 1.]
  [1. 1. 1. 1.]]

 [[1. 1. 1. 1.]
  [1. 1. 1. 1.]]

 [[1. 1. 1. 1.]
  [1. 1. 1. 1.]]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[[1. 1. 1. 1.]<br>  [1. 1. 1. 1.]]<br><br> [[1. 1. 1. 1.]<br>  [1. 1. 1. 1.]]<br><br> [[1. 1. 1. 1.]<br>  [1. 1. 1. 1.]]]|

**c.** Inisialisasikan sebuah array 1D berukuran 5 dengan nilai acak.

In [8]:
arr = np.random.rand(5)
print(arr)

[0.39361873 0.54505132 0.10973774 0.48289794 0.87162412]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[0.75565203 0.36572656 0.31421755 0.37958106 0.72299638]<br>*Catatan: Angka akan tergenerate secara random*|

**d.** Inisialisasikan sebuah array 1D dengan range nilai dari 7 hingga 17.

In [9]:
# jawab disini
arr = np.arange(7, 18)
print(arr)

[ 7  8  9 10 11 12 13 14 15 16 17]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[ 7 8 9 10 11 12 13 14 15 16 17]|

**e.** Inisialisasikan sebuah array 1D yang memiliki nilai yang terdistribusi merata antara 0 hingga 1 dengan 5 elemen </br> Tip: gunakan .linspace()

In [10]:
#jawab disini
arr = np.linspace(0, 1, 5)
print(arr)

[0.   0.25 0.5  0.75 1.  ]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[0. 0.25 0.5 0.75 1. ]|

**f.** Buatlah array 2D berukuran 2x3 dengan semua elemen diisi dengan angka 7

In [11]:
#jawab disini
arr = np.full((2, 3), 7)
print(arr)

[[7 7 7]
 [7 7 7]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[7 7 7]<br> [7 7 7]]|

**g.** Buatlah sebuah array 1D dengan 4 nilai acak antara 0 dan 1

In [12]:
#jawab disini
arr = np.random.rand(4)
print(arr)

[0.11236331 0.17651211 0.31410235 0.32964565]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[0.67039964 0.38240816 0.9551951  0.07901245]<br>*Catatan: Angka akan tergenerate secara random*|

**h.** Buatlah array 2D berukuran 3x3 yang berisi bilangan acak bulat antara 1 dan 10

In [13]:
#jawab disini
arr = np.random.randint(1, 11, (3, 3))
print(arr)

[[10 10 10]
 [ 3  5  7]
 [ 7  1  6]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[5 2 9] <br>[1 8 3] <br>[7 6 4]]<br>*Catatan: Angka akan tergenerate secara random*|

**i.** Buatlah array 2D berukuran 3x3 yang isinya merupakan menghasilkan nilai dari perkalian baris dan kolomnya

In [14]:
# Buat array 3x3 di mana nilai tiap elemen = perkalian indeks baris × indeks kolom
i = np.arange(3, dtype=float).reshape(3, 1)
j = np.arange(3, dtype=float)
arr = i * j

print(arr)

[[0. 0. 0.]
 [0. 1. 2.]
 [0. 2. 4.]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[0. 0. 0.] <br>[0. 1. 2.] <br>[0. 2. 4.]]|

### 4. Penambahan Elemen Array (`5`)

**a.** Anda bekerja dengan data mahasiswa yang disimpan dalam array NumPy. Data ini terdiri dari nama-nama mahasiswa dari dua kelas berbeda. Tugas Anda adalah menggabungkan data dari kedua kelas tersebut ke dalam satu array.

In [15]:
# Data dari kelas A dan kelas B
kelas_A = np.array(['Murid A', 'Murid A2', 'Murid A3'])
kelas_B = np.array(['Murid B', 'Murid B1', 'Murid B2'])

# Gabungkan kedua array
hasil = np.concatenate((kelas_A, kelas_B))

# Tampilkan array hasil gabungan
print(hasil)

['Murid A' 'Murid A2' 'Murid A3' 'Murid B' 'Murid B1' 'Murid B2']


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |['Murid A' 'Murid A2' 'Murid A3' 'Murid B' 'Murid B1' 'Murid B2']|

**b.** Dua matakuliah memerlukan penggabungan data nilai untuk analisis lebih lanjut. Nilai kedua matkul tersebut disimpan dalam array terpisah. Gabungkan kedua array ini menjadi satu. **Gunakan sintaks penggabungan array yang berbeda dengan soal sebelumnya.**

In [16]:
# Nilai matkul A dan B
nilai_A = np.array([75, 80, 85])
nilai_B = np.array([70, 90, 95])

# Gabungkan kedua array (menggunakan np.hstack, berbeda dari soal sebelumnya)
hasil = np.hstack((nilai_A, nilai_B))

# Tampilkan array hasil gabungan
print(hasil)

[75 80 85 70 90 95]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[75 80 85 70 90 95]|

**c.** Anda diminta untuk menambahkan nilai ujian tengah semester (UTS) pada posisi tertentu dalam array nilai mahasiswa.

In [17]:
# Nilai akhir mahasiswa
nilai = np.array([85, 90, 95])

# Sisipkan nilai UTS (88) di posisi ke-2
nilai = np.insert(nilai, 1, 88)

# Tampilkan array setelah penyisipan nilai UTS
print(nilai)

[85 88 90 95]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[85 88 90 95]|

**d.** Anda diminta untuk menggabungkan dua array nilai ke dalam sebuah array 2D untuk mempermudah analisis lebih lanjut. Usahakan gunakan sintaks penggabungan array yang berbeda dengan soal sebelumnya.

In [18]:
# Nilai dari dua kelas
nilai_kelas_A = np.array([80, 85, 90])
nilai_kelas_B = np.array([75, 88, 92])

# Gabungkan menjadi array 2D
nilai_gabungan = np.vstack((nilai_kelas_A, nilai_kelas_B))

print(nilai_gabungan)

[[80 85 90]
 [75 88 92]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[80 85 90]<br> [75 88 92]]|

### 5. Penghapusan Elemen Array (`10`)

**a.** Seorang peneliti ingin membersihkan data yang mengandung elemen salah dalam sebuah array 1D yang berisi nilai eksperimen. Hapus elemen sesuai dengan index input si peneliti.

In [19]:
# Array berisi nilai eksperimen
nilai_eksperimen = np.array([10, 15, 18, 22])

# Hapus elemen di indeks yang sesuai
indeks_hapus = int(input("index yang ingin dihapus: "))
nilai_eksperimen = np.delete(nilai_eksperimen, indeks_hapus)

# Tampilkan array setelah penghapusan elemen
print(nilai_eksperimen)

index yang ingin dihapus: 1
[10 18 22]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|1|[10 18 22]|
|3|[10 15 18]|

**b.** Anda bekerja dengan data hasil tes dari beberapa kelompok siswa yang disimpan dalam array 2D. Namun, data dari kelompok ke-3 dan semua siswa pertama harus dihapus karena salah pencatatan. Hapus baris ketiga dan kolom pertama dari array tersebut.

In [20]:
# Data hasil tes (dalam 2D array)
hasil_tes = np.array([[85, 90, 88],
                      [78, 82, 84],
                      [93, 89, 92],
                      [75, 80, 85]])

print("Masukkan nomor kelompok dan urutan siswa, dengan dipisahkan oleh koma (,)")

# Simpan nomor kelompok (baris) dan urutan siswa (kolom) dalam list
target_hapus = input("nomor kelompok dan urutan siswa yang ingin dihapus: ").split(",")

indeks_baris = int(target_hapus[0]) - 1
indeks_kolom = int(target_hapus[1]) - 1

# Hapus data dari sesuai dengan target hapus
hasil_tes = np.delete(hasil_tes, indeks_baris, axis=0)
hasil_tes = np.delete(hasil_tes, indeks_kolom, axis=1)

#Tampilkan array setelah penghapusan data
print(hasil_tes)

Masukkan nomor kelompok dan urutan siswa, dengan dipisahkan oleh koma (,)
nomor kelompok dan urutan siswa yang ingin dihapus: 3,1
[[90 88]
 [82 84]
 [80 85]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|3,1|[[90 88]<br>[82 84]<br>[80 85]]|
|4,2|[[85 88]<br>[78 84]<br>[93 92]]|

**c.** Seorang ilmuwan ingin menghapus semua elemen yang berada di bawah batas nilai dalam array yang berisi data eksperimen.

In [21]:
# Data eksperimen
data = np.array([10, 3, 15, 7, 22, 17], dtype=float)

# Menerima input batas nilai
batas_nilai = int(input("masukkan batas nilai: "))

# Menghapus elemen dibawah batas nilai dari array
data = data[data >= batas_nilai]

#Tampilkan array setelah penghapusan elemen
print(data)

masukkan batas nilai: 15
[15. 22. 17.]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|15|[15. 22. 17.]|
|8|[10. 15. 22. 17.]|

### 6. Pengurutan Elemen Array (`5`)

**a.** Seorang petugas inventaris ingin menyusun harga barang-barang yang sudah diinput ke dalam sistem.

In [22]:
# Data harga barang
harga_barang = np.array([35000, 25000, 50000, 45000, 30000])

# Urutkan harga barang
harga_barang = np.sort(harga_barang)

# Tampilkan array setelah pengurutan
print(harga_barang)

[25000 30000 35000 45000 50000]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[25000 30000 35000 45000 50000]|

**b.** Anda memiliki data hasil tes siswa dalam array 2D, di mana setiap baris adalah skor dari satu siswa.

In [23]:
# Data hasil tes siswa
skor_siswa = np.array([[85, 78, 90], [92, 89, 85], [76, 84, 80]])

# Urutkan skor setiap siswa
skor_siswa = np.sort(skor_siswa, axis=1)

# Tampilkan skor yang sudah diurutkan
print(skor_siswa)

[[78 85 90]
 [85 89 92]
 [76 80 84]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[78 85 90]<br> [85 89 92]<br> [76 80 84]]|

**c.** Tim peneliti ingin mengurutkan semua data dalam array multidimensi berdasarkan nilainya (secara keseluruhan, bukan hanya per baris atau kolom).

In [24]:
# Data eksperimen
data = np.array([[9, 4, 6], [3, 7, 1]])

# Urutkan semua elemen menjadi 1 array
data_flat = data.flatten()

# Tampilkan skor yang sudah diurutkan
print(np.sort(data_flat))

[1 3 4 6 7 9]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[1 3 4 6 7 9]|

### 7. Indexing dan Slicing Array (`10`)

**a.** Seorang peneliti memiliki data suhu harian selama satu bulan dalam bentuk array (30 hari).</br> Peneliti ingin mengetahui suhu pada hari ke-5, hari ke-10, hari ke-25, dan hari ke-30. Setelah mendapatkan suhu-suhu tersebut, lakukan perhitungan rata-rata.

In [25]:
# Data suhu selama 30 hari dengan range suhu 20 sampai 40 (buat menggunakan inisialisasi random integer)
suhu = np.random.randint(20, 41, 30)

# Ambil suhu pada hari ke-5, 10, 25, dan 30
suhu_hari_tertentu = suhu[[4, 9, 24, 29]]

# Hitung rata-rata suhu
rata_rata_suhu = np.mean(suhu_hari_tertentu)

print(f"Suhu hari tertentu: {suhu_hari_tertentu}")
print(f"Rata-rata: {rata_rata_suhu}")

Suhu hari tertentu: [37 27 30 31]
Rata-rata: 31.25


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Suhu hari tertentu: [31 31 25 25]<br>Rata-rata: 28.0<br>*Catatan: Angka akan tergenerate secara random*|

**b.** Seorang ilmuwan mencatat jumlah bakteri selama 10 hari dalam sebuah array. Namun, ia hanya ingin mengambil data pertumbuhan bakteri setiap 2 hari (hari ke-1, 3, 5, dan seterusnya) dan menghitung total pertumbuhannya.

In [26]:
# Data pertumbuhan bakteri selama 10 hari
pertumbuhan_bakteri = np.array([100, 120, 240, 130, 150, 110, 220, 175, 160, 200])

# Ambil data setiap 2 hari
pertumbuhan_bakteri_setiap_2_hari = pertumbuhan_bakteri[::2]

# Hitung total pertumbuhan
total_pertumbuhan = np.sum(pertumbuhan_bakteri_setiap_2_hari)

# Tampilkan hasil
print(f"Data bakteri: {pertumbuhan_bakteri_setiap_2_hari}")
print(f"Total pertumbuhan: {total_pertumbuhan}")

Data bakteri: [100 240 150 220 160]
Total pertumbuhan: 870


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data bakteri: [100 240 150 220 160]<br>Total pertumbuhan: 870|

**c.** Anda memiliki array berisi data stok barang di gudang. Anda ingin menghapus semua barang yang stoknya kurang dari 50 unit, kemudian hitung total stok barang yang tersisa. Gunakan boolean indexing untuk menyaring data dan hitung totalnya.

In [27]:
# Data stok barang
stok_barang = np.array([41, 51, 75, 57, 92, 13, 35, 64, 24, 66])

# Filter barang dengan stok >= 50
barang_tersedia = stok_barang[stok_barang >= 50]

# Hitung total stok tersisa
total_stok_tersedia = np.sum(barang_tersedia)

# Tampilkan hasil sesuai contoh output
print(f"Stok barang tersisa: {barang_tersedia}")
print(f"Total stok tersisa: {total_stok_tersedia}")

Stok barang tersisa: [51 75 57 92 64 66]
Total stok tersisa: 405


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Stok barang tersisa: [51 75 57 92 64 66]<br>Total stok tersisa: 405|

### 8. Penyalinan (Copy) Array (`5`)

**a.** Seorang ilmuwan data ingin melakukan analisis terhadap data eksperimen dengan mengubah data aslinya. Buat salinan "shallow copy" dari array data asli, lalu ubah elemen tertentu pada indeks tertentu dari data yang telah disalin.

In [28]:
# Data eksperimen asli
data_asli = np.array([10, 15, 20, 25, 30])

# Buat shallow copy
data_copy = data_asli

# Masukkan indeks dan data baru
informasi_perubahan = input("masukkan index yang ingin diubah dan data baru: ").split(",")

# Ubah data
data_copy[int(informasi_perubahan[0])] = int(informasi_perubahan[1])

# Tampilkan hasil
print("data copy:", data_copy)
print("data asli:", data_asli)

masukkan index yang ingin diubah dan data baru: 1,100
data copy: [ 10 100  20  25  30]
data asli: [ 10 100  20  25  30]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|1,100|Data Copy: [ 10 100  20  25  30]<br>Data Asli: [ 10 100  20  25  30]|

**b.** Seorang peneliti ingin melakukan eksperimen dengan data cuaca, dan memastikan bahwa perubahan data baru tidak mempengaruhi data aslinya. Buat "deep copy" dari array cuaca dan ubah data salinan tersebut.

In [29]:
# Data cuaca asli
data_cuaca = np.array([[30, 32, 35], [28, 27, 26], [33, 31, 29]])

# Buat deep copy dari data_cuaca
data_cuaca_copy = data_cuaca.copy()

# Ubah elemen di baris ke-2, kolom ke-3 data_cuaca_copy menjadi 40
data_cuaca_copy[1, 2] = 40

# Tampilkan data sesuai contoh output
print("data cuaca copy:\n", data_cuaca_copy)
print("data cuaca asli:\n", data_cuaca)

data cuaca copy:
 [[30 32 35]
 [28 27 40]
 [33 31 29]]
data cuaca asli:
 [[30 32 35]
 [28 27 26]
 [33 31 29]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data Cuaca Copy:<br> [[30 32 35]<br> [28 27 40]<br>[33 31 29]]<br>Data Cuaca Asli:<br> [[30 32 35]<br> [28 27 26]<br> [33 31 29]]|

### 9. Operasi Dasar Array (`5`)

**a.** Seorang guru ingin mengetahui nilai tertinggi yang diperoleh siswa dari hasil ujian matematika. </br> Tambahkan juga kode untuk menghitung selisih antara nilai maksimum dan nilai rata-rata ujian.

In [30]:
# Data nilai ujian siswa dalam array
nilai_ujian = np.array([75, 80, 90, 85, 60])

# Temukan nilai maksimum
nilai_maksimum = np.max(nilai_ujian)

# Hitung nilai rata-rata
nilai_rata_rata = np.mean(nilai_ujian)

# Hitung selisih antara nilai maksimum dan rata-rata
nilai_selisih = nilai_maksimum - nilai_rata_rata

# Tampilkan nilai maksimum dan selisihnya
print(f"Nilai Maksimum: {nilai_maksimum}")
print(f"Selisih Maksimum dan Rata-rata: {nilai_selisih}")

Nilai Maksimum: 90
Selisih Maksimum dan Rata-rata: 12.0


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Nilai Maksimum: 90<br> Selisih Maksimum dan Rata-rata: 12.0|

**b.** Seorang pelatih renang ingin mengetahui waktu renang tercepat dari setiap atlet. Hitung juga perbedaan waktu antara waktu tercepat dan waktu rata-rata.

In [31]:
# Data waktu renang atlet
waktu_renang = np.array([50.5, 52.0, 48.7, 49.8, 51.2])

# Waktu tercepat
waktu_tercepat = np.min(waktu_renang)

# Rata-rata
waktu_rata_rata = np.mean(waktu_renang)

# Selisih (dibulatkan 1 angka desimal)
selisih = round(waktu_tercepat - waktu_rata_rata, 0)

print(f"Waktu Tercepat: {waktu_tercepat}")
print(f"Selisih Waktu Tercepat dan Rata-rata: {selisih}")

Waktu Tercepat: 48.7
Selisih Waktu Tercepat dan Rata-rata: -2.0


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Waktu Tercepat: 48.7<br> Selisih Waktu Tercepat dan Rata-rata: -2.0|

**c.** Dalam sebuah kompetisi, seorang juri ingin mengetahui total skor yang diperoleh tim A selama lima putaran pertandingan. Tampilkan rata-rata skor per putaran.

In [32]:
# Data skor tim A dalam 5 putaran
skor_tim_A = np.array([85, 78, 90, 88, 76])

# Hitung total skor
total_skor = np.sum(skor_tim_A)

# Hitung rata-rata skor per putaran (menggunakan np.mean)
rata_rata_skor = np.mean(skor_tim_A)

# Tampilkan total skor dan rata-ratanya
print(f"Total Skor: {total_skor}")
print(f"Rata-rata Skor per Putaran: {rata_rata_skor}")

Total Skor: 417
Rata-rata Skor per Putaran: 83.4


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Total Skor: 417<br> Rata-rata Skor per Putaran: 83.4|

### 10. Transpose dan Reshape Array (`15`)

**a.** Seorang dosen ingin menampilkan data mahasiswa dalam format yang berbeda. Data saat ini disusun dalam bentuk baris, di mana setiap baris merepresentasikan mahasiswa. **Anda diminta untuk membuat array tersebut sehingga data mahasiswa ditampilkan per kolom.**

In [33]:
# Data mahasiswa (Nama, NIM, Jurusan)
data_mahasiswa = np.array([['Farah', 'Tiara', 'Imelda'], ['123', '456', '789'], ['IF', 'DS', 'RPL']])

# Ubah data_mahasiswa
data_mahasiswa = data_mahasiswa.T

# Tampilkan data mahasiswa
print("Data mahasiswa:")
print(data_mahasiswa)

Data mahasiswa:
[['Farah' '123' 'IF']
 ['Tiara' '456' 'DS']
 ['Imelda' '789' 'RPL']]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data mahasiswa:<br> [['Farah' '123' 'IF']<br> ['Tiara' '456' 'DS']<br> ['Imelda' '789' 'RPL']]|

In [34]:
# Data nilai mahasiswa
nilai = np.array([80, 90, 85, 78, 92, 88])

# Bentuk ulang array nilai menjadi 2D dengan 3 kolom
nilai_2d = nilai.reshape(2, 3)

# Tampilkan array nilai
print("Array nilai:")
print(nilai_2d)

Array nilai:
[[80 90 85]
 [78 92 88]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Array nilai:<br> [[80 90 85]<br> [78 92 88]]|

**c.** Sebuah perusahaan ingin menganalisis data penjualan dari berbagai cabang. Data saat ini ada dalam bentuk array 1D. Ubah array tersebut menjadi bentuk 2D dan transposisikan hasilnya agar data cabang dan produk lebih mudah dibaca.

In [35]:
# Data penjualan
penjualan = np.array([120, 150, 180, 100, 130, 170])

# Ubah array
penjualan_2d = penjualan.reshape(3, 2).T

# Tampilkan data penjualan
print("Data penjualan:")
print(penjualan_2d)

Data penjualan:
[[120 180 130]
 [150 100 170]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data penjualan:<br> [[120 180 130]<br>  [150 100 170]]|

### 11. Soal Cerita (`20`)

Seorang manajer toko retail sedang menganalisis tingkat kepuasan pelanggan (skala 5 - 10) dari tiga cabang toko selama 8 minggu. Data kepuasan disimpan dalam sebuah array yang merepresentasikan rating kepuasan dari tiga cabang (kolom) di setiap minggu (baris). Manajer hanya tertarik pada minggu-minggu di mana rata-rata kepuasan ketiga cabang lebih dari atau sama dengan 8. Mereka ingin menganalisis lebih lanjut dengan menyalin data minggu tersebut, melakukan transpose pada data untuk melihat statistik dari perspektif cabang, dan menghitung rata-rata kepuasan tiap cabang. Berikut adalah langkah-langkah yang perlu dilakukan:

- Ambil subset data dari minggu-minggu yang dimana rata-rata kepuasan 3 cabang pada minggu yang sama memiliki nilai lebih dari atau sama dengan 8
- Buat salinan data untuk dianalisis lebih lanjut
- *Transpose array* hasil salinan statistik
- Ambil rata-rata kepuasan tiap cabang pada minggu (baris) yang terdapat pada subset dari data kepuasan

In [36]:
# Buat array yang merepresentasikan data kepuasan 8 minggu dari 3 cabang (8x3)
data_kepuasan = np.array([
    [9, 7, 8],
    [8, 9, 10],
    [7, 6, 8],
    [10, 8, 7],
    [6, 7, 5],
    [8, 8, 9],
    [5, 6, 7],
    [9, 10, 8]
])

# You're on Your Own, Kid. Good Luck :)

# Tampilkan data asli
print("Data Kepuasan Asli:")
print(data_kepuasan)

# Hitung rata-rata tiap minggu
rata_mingguan = np.mean(data_kepuasan, axis=1)

# Ambil minggu dengan rata-rata >= 8
subset = data_kepuasan[rata_mingguan >= 8]

print("Subset Data Kepuasan:")
print(subset)

# Buat salinan data
subset_copy = subset.copy()

# Transpose
transpose_data = subset_copy.T

print("Data Setelah Di-transpose:")
print(transpose_data)

# Hitung rata-rata tiap cabang
rata_cabang = np.mean(transpose_data, axis=1)

print("Rata - Rata Kepuasan Cabang A: ", rata_cabang[0])
print("Rata - Rata Kepuasan Cabang B: ", rata_cabang[1])
print("Rata - Rata Kepuasan Cabang C: ", rata_cabang[2])

Data Kepuasan Asli:
[[ 9  7  8]
 [ 8  9 10]
 [ 7  6  8]
 [10  8  7]
 [ 6  7  5]
 [ 8  8  9]
 [ 5  6  7]
 [ 9 10  8]]
Subset Data Kepuasan:
[[ 9  7  8]
 [ 8  9 10]
 [10  8  7]
 [ 8  8  9]
 [ 9 10  8]]
Data Setelah Di-transpose:
[[ 9  8 10  8  9]
 [ 7  9  8  8 10]
 [ 8 10  7  9  8]]
Rata - Rata Kepuasan Cabang A:  8.8
Rata - Rata Kepuasan Cabang B:  8.4
Rata - Rata Kepuasan Cabang C:  8.4


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data Kepuasan Asli:<br>[[ 9  7  8]<br>[ 8  9 10]<br>[ 7  6  8]<br>[10  8  7]<br>[ 6  7  5]<br>[ 8  8  9]<br>[ 5  6  7]<br>[ 9 10  8]]<br>Subset Data Kepuasan:<br>[[ 9  7  8]<br>[ 8  9 10]<br>[10  8  7]<br>[ 8  8  9]<br>[ 9 10  8]]<br>Data Setelah Di-transpose:<br>[[ 9  8 10  8  9]<br>[ 7  9  8  8 10]<br>[ 8 10  7  9  8]]<br>Rata - Rata Kepuasan Cabang A:  8.8<br>Rata - Rata Kepuasan Cabang B:  8.4<br>Rata - Rata Kepuasan Cabang C:  8.4|